File: zegl-go/day1.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func day1part1(input string) int {
	instructions := strings.Split(input, ", ")

	// 0 = North
	// 1 = East
	// 2 = South
	// 3 = West
	currentlyFacing := 0
	currentX := 0
	currentY := 0

	for _, ins := range instructions {
		// Left or right
		lorr := ins[0]

		if lorr == 'R' {
			// Turn right
			currentlyFacing++
		} else {
			// Turn left
			currentlyFacing--
		}

		// Correct over and underflows
		if currentlyFacing < 0 {
			currentlyFacing = 3
		}

		if currentlyFacing > 3 {
			currentlyFacing = 0
		}

		steps, _ := strconv.Atoi(ins[1:])

		switch currentlyFacing {
		case 0:
			currentY += steps
		case 1:
			currentX += steps
		case 2:
			currentY -= steps
		case 3:
			currentX -= steps
		}
	}

	return int(math.Abs(float64(currentX)) + math.Abs(float64(currentY)))
}

func day1part2(input string) int {
	instructions := strings.Split(input, ", ")

	// 0 = North
	// 1 = East
	// 2 = South
	// 3 = West
	currentlyFacing := 0
	currentX := 0
	currentY := 0

	visited := make(map[string]bool)

	for _, ins := range instructions {
		// Left or right
		lorr := ins[0]

		if lorr == 'R' {
			// Turn right
			currentlyFacing++
		} else {
			// Turn left
			currentlyFacing--
		}

		// Correct over and underflows
		if currentlyFacing < 0 {
			currentlyFacing = 3
		}

		if currentlyFacing > 3 {
			currentlyFacing = 0
		}

		steps, _ := strconv.Atoi(ins[1:])

		for i := 0; i < steps; i++ {

			switch currentlyFacing {
			case 0:
				currentY++
			case 1:
				currentX++
			case 2:
				currentY--
			case 3:
				currentX--
			}

			key := strconv.Itoa(currentX) + "," + strconv.Itoa(currentY)

			// We have already visited this location
			if _, ok := visited[key]; ok {
				return int(math.Abs(float64(currentX)) + math.Abs(float64(currentY)))
			}

			visited[key] = true

		}
	}

	fmt.Println("Did not find duplicate?")
	return 0
}

File: plilja-python3/day1/day1.py
def walk(inp):
    x = 0
    y = 0
    dx = 0
    dy = -1 # face north initially
    r = [(0, 0)]
    for s in inp.replace(',', '').split():
        if s[0] == 'L':
            dx, dy = dy, -dx
        else:
            dx, dy = -dy, dx
        steps = int(s[1:])
        for i in range(0, steps):
            x += dx
            y += dy
            r += [(x, y)]
    return r

def step1(inp):
    (endx, endy) = walk(inp)[-1]
    return abs(endx) + abs(endy)

def step2(inp):
    w = walk(inp)
    visited = set()
    for w in walk(inp):
        if w in visited:
            return abs(w[0]) + abs(w[1])
        visited |= {w}
    return -1

inp = input()
print(step1(inp))
print(step2(inp))
